src/openms/source/ANALYSIS/ID/FIAMSDataProcessor.cpp

```diff
@@ -0,0 +1,422 @@
+    dir_input_(dir_input),
+    dir_output_(dir_output),
+    resolution_(resolution),
+    polarity_(polarity),
+    min_mz_(min_mz),
+    max_mz_(max_mz),
+    bin_step_(bin_step),
+    db_mapping_(db_mapping),
+    db_struct_(db_struct),
+    positive_adducts_(positive_adducts),
+    negative_adducts_(negative_adducts),
+    store_progress_(store_progress),
+    mzs_(),
+    bin_sizes_()
+  {
+  size_t n_bins = static_cast<int> (max_mz_ - min_mz_) / bin_step_;
+  for (size_t i = 0; i < n_bins; i++) {
+      mzs_.push_back(i*bin_step_);
+      bin_sizes_.push_back(mzs_[i] / (resolution_*4.0));
+  }
+  loadExperiment_();
+}
+
+/// default destructor
+FIAMSDataProcessor::~FIAMSDataProcessor() {
+}
+
+/// copy constructor
+FIAMSDataProcessor::FIAMSDataProcessor(const FIAMSDataProcessor& source) :
+  filename_(source.filename_),
+  dir_input_(source.dir_input_),
+  dir_output_(source.dir_output_),
+  resolution_(source.resolution_),
+  polarity_(source.polarity_),
+  min_mz_(source.min_mz_),
+  max_mz_(source.max_mz_),
+  bin_step_(source.bin_step_),
+  db_mapping_(source.db_mapping_),
+  db_struct_(source.db_struct_),
+  positive_adducts_(source.positive_adducts_),
+  negative_adducts_(source.negative_adducts_),
+  mzs_(source.mzs_),
+  store_progress_(source.store_progress_),
+  bin_sizes_(source.bin_sizes_)
+  {
+  }
+
+/// assignment operator
+FIAMSDataProcessor& FIAMSDataProcessor::operator=(const FIAMSDataProcessor& rhs) {
+  if (this == &rhs) return *this;
+  filename_ = rhs.filename_;
+  dir_input_ = rhs.dir_input_;
+  dir_output_ = rhs.dir_output_;
+  resolution_ = rhs.resolution_;
+  polarity_ = rhs.polarity_;
+  min_mz_ = rhs.min_mz_;
+  max_mz_ = rhs.max_mz_;
+  bin_step_ = rhs.bin_step_;
+  db_mapping_ = rhs.db_mapping_;
+  db_struct_ = rhs.db_struct_;
+  positive_adducts_ = rhs.positive_adducts_;
+  negative_adducts_ = rhs.negative_adducts_;
+  store_progress_ = rhs.store_progress_;
+  mzs_ = rhs.mzs_;
+  bin_sizes_ = rhs.bin_sizes_;
+  return *this;
+}
+
+void FIAMSDataProcessor::cutForTime(const MSExperiment & experiment, vector<MSSpectrum> & output, float n_seconds) {
+    for (auto s : experiment.getSpectra()) {
+        if (s.getRT() < n_seconds) output.push_back(s);
+    }
+}
+
+MSSpectrum FIAMSDataProcessor::mergeAlongTime(
+  const std::vector<MSSpectrum> & input
+  ) {
+    MSSpectrum output;
+    for (size_t i = 1; i < mzs_.size() - 1; i++) {
+        OpenMS::MSSpectrum full_spectrum = OpenMS::SpectrumAddition::addUpSpectra(
+            input, bin_sizes_[i], false
+        );
+        for (auto it = full_spectrum.begin(); it != full_spectrum.end(); ++it) {
+            if (it->getMZ() > mzs_[i+1]) break;
+            if (it->getMZ() >= mzs_[i]) output.push_back(*it);
+        }
+    }
+    output.sortByPosition();
+    return output;
+}
+
+MSSpectrum FIAMSDataProcessor::extractPeaks(const MSSpectrum & input) {
+  MSSpectrum spectrum(input);
+
+  SavitzkyGolayFilter filter;
+  filter.filter(spectrum);
```

Comment:
SGolay filter assumes equally spaced points, is that the case here? Will this always work well? Also, would you need a way to control parameters?

Finally, this will be pretty slow since constructing an object derived from `DefaultParamHandler` for each spectrum is rather slow...